// ng-nice-select/esm5/lib/ng-nice-select.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef } from '@angular/core';
import { Directive } from '@angular/core';
import 'jquery-nice-select';
var NgNiceSelectDirective = /** @class */ (function () {
    function NgNiceSelectDirective(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        $(this.el.nativeElement).niceSelect();
        this.attachAngularChange();
        this.attachObserver();
    };
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.attachAngularChange = /**
     * @return {?}
     */
    function () {
        var _this = this;
        $(this.el.nativeElement).one('change', function () {
            /** @type {?} */
            var customEvent = document.createEvent('Event');
            customEvent.initEvent('change', true, true);
            _this.el.nativeElement.dispatchEvent(customEvent);
            _this.attachAngularChange();
        });
    };
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.attachObserver = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.observer = new MutationObserver(function (mutations) {
            mutations.forEach(function () { return _this.onChanges(); });
        });
        this.observer.observe(this.el.nativeElement, {
            attributes: true,
            childList: true,
            characterData: true
        });
    };
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.detachObserver = /**
     * @return {?}
     */
    function () {
        this.observer.disconnect();
    };
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.onChanges = /**
     * @return {?}
     */
    function () {
        this.detachObserver();
        $(this.el.nativeElement).niceSelect('update');
        this.attachObserver();
    };
    /**
     * @return {?}
     */
    NgNiceSelectDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.detachObserver();
        $(this.el.nativeElement).niceSelect('destroy');
    };
    NgNiceSelectDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[nice-select]',
                },] }
    ];
    /** @nocollapse */
    NgNiceSelectDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return NgNiceSelectDirective;
}());
export { NgNiceSelectDirective };
if (false) {
    /** @type {?} */
    NgNiceSelectDirective.prototype.observer;
    /**
     * @type {?}
     * @private
     */
    NgNiceSelectDirective.prototype.el;
}
